Generated java tests fails when a field has unique validation, entity has filter and is required

##### **Overview of the issue**

Generated java tests fails when a field has unique validation and filter:

```
[ERROR] Errors: 
[ERROR]   EmployeeResourceIT.getAllEmployeesByLocationIsEqualToSomething:324 » Persistence
[INFO] 
[ERROR] Tests run: 141, Failures: 0, Errors: 1, Skipped: 0
```
```
[ERROR] Tests run: 20, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.312 s <<< FAILURE! - in com.mycompany.myapp.web.rest.EmployeeResourceIT
[ERROR] getAllEmployeesByLocationIsEqualToSomething  Time elapsed: 0.137 s  <<< ERROR!
javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at com.mycompany.myapp.web.rest.EmployeeResourceIT.getAllEmployeesByLocationIsEqualToSomething(EmployeeResourceIT.java:324)
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at com.mycompany.myapp.web.rest.EmployeeResourceIT.getAllEmployeesByLocationIsEqualToSomething(EmployeeResourceIT.java:324)
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: 
Unique index or primary key violation: "PUBLIC.UX_LOCATION_NAME_INDEX_9 ON PUBLIC.LOCATION(NAME) VALUES 4"; SQL statement:
insert into location (id, description, name) values (null, ?, ?) [23505-199]
	at com.mycompany.myapp.web.rest.EmployeeResourceIT.getAllEmployeesByLocationIsEqualToSomething(EmployeeResourceIT.java:324)
```


##### **Motivation for or Use Case**

The generated tests should be passing.

##### **Reproduce the error**

Create a new project with default settings `jhipster`, import JDL `jhipster import-jdl bug.jdl`, run `./mvnw clean install`.

##### **Related issues**

It was also happening in #9563 , but I hadn't identified the cause and it was also happening with other bugs.

##### **Suggest a Fix**

The createEntity shouldn't create entities that have the same value for the field if it is unique, as it could be created twice by other entities that require it.

Another simple solution is not to create a new entity if it was already created. I will try to make a PR to fix it.

##### **JHipster Version(s)**

```
teste-6-simples@0.0.0 /home/gabriel/projetos/teste6simples
└── generator-jhipster@6.0.0 

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.mycompany.myapp",
      "nativeLanguage": "pt-br"
    },
    "jhipsterVersion": "6.0.0",
    "applicationType": "monolith",
    "baseName": "teste6simples",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "clientFramework": "angularX",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "useSass": true,
    "clientPackageManager": "npm",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "pt-br",
    "languages": ["pt-br"]
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
```
entity Employee {
  name String required,
  age Integer required
}
entity Location {
  name String required unique,
  description String
}
relationship ManyToOne {
  Employee{location(name) required} to Location
}
filter Employee
```

##### **Environment and Tools**

<details>
java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)

git version 2.11.0

node: v10.15.3

npm: 6.4.1

yeoman: 2.0.6

yarn: 1.15.2

Docker version 18.09.5, build e8ff056dbc

docker-compose version 1.24.0, build 0aa59064
</details>

##### **Browsers and Operating System**

Debian 9

-   [x] Checking this box is mandatory (this is just to show you read everything)

The additions I made to the app.jdl catches the behaviour: in the current version these are the errors of the [app.jdl updated](https://github.com/Gabrui/generator-jhipster/commit/6c69ff6c0d081a321a26c86e549293fa3d4f43b0):

```
[ERROR] getEmployee  Time elapsed: 0.087 s  <<< ERROR!
javax.persistence.PersistenceException: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.createEntity(EmployeeResourceIT.java:143)
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.initTest(EmployeeResourceIT.java:152)
Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.createEntity(EmployeeResourceIT.java:143)
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.initTest(EmployeeResourceIT.java:152)
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: 
Unique index or primary key violation: "PUBLIC.UX_IDENTIFIER_NAME_INDEX_B ON PUBLIC.IDENTIFIER(NAME) VALUES 23"; SQL statement:
insert into identifier (id, name) values (null, ?) [23505-199]
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.createEntity(EmployeeResourceIT.java:143)
	at io.github.jhipster.sample.web.rest.EmployeeResourceIT.initTest(EmployeeResourceIT.java:152)


[ERROR]   EmployeeResourceIT.initTest:152->createEntity:143 » Persistence org.hibernate....
[INFO] 
[ERROR] Tests run: 335, Failures: 0, Errors: 46, Skipped: 0
```

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -411,7 +411,7 @@ _%>
 
 <%_ ['DEFAULT_', 'UPDATED_'].forEach((fieldStatus) => { _%>
     /**
-     * Create an <% if (fieldStatus === 'UPDATED_') { %>updated<% } %> entity for this test.
+     * Create an <% if (fieldStatus === 'UPDATED_') { %>updated<% } %>entity for this test.
      *
      * This is a static method, as tests for other entities might also need it,
      * if they test an entity which requires the current entity.

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -411,7 +411,7 @@ _%>
 
 <%_ ['DEFAULT_', 'UPDATED_'].forEach((fieldStatus) => { _%>
     /**
-     * Create an <% if (fieldStatus === 'UPDATED_') { %>updated<% } %>entity for this test.
+     * Create an <% if (fieldStatus === 'UPDATED_') { %>updated <% } %>entity for this test.
      *
      * This is a static method, as tests for other entities might also need it,
      * if they test an entity which requires the current entity.

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -409,23 +409,24 @@ _%>
             .setValidator(validator).build();
     }
 
+<%_ ['DEFAULT_', 'UPDATED_'].forEach((fieldStatus) => { _%>
     /**
-     * Create an entity for this test.
+     * Create an <% if (fieldStatus === 'UPDATED_') { %>updated<% } %> entity for this test.
      *
      * This is a static method, as tests for other entities might also need it,
      * if they test an entity which requires the current entity.
      */
-    public static <%= asEntity(entityClass) %> createEntity(<% if (databaseType === 'sql') { %>EntityManager em<% } %>) {
+    public static <%= asEntity(entityClass) %> create<% if (fieldStatus === 'UPDATED_') { _%>Updated<%_ } %>Entity(<% if (databaseType === 'sql') { %>EntityManager em<% } %>) {
         <%_ if (fluentMethods) { _%>
         <%= asEntity(entityClass) %> <%= asEntity(entityInstance) %> = new <%= asEntity(entityClass) %>()<% for (idx in fields) { %>
-            .<%= fields[idx].fieldName %>(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase()%>)<% if ((fields[idx].fieldType === 'byte[]' || fields[idx].fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent !== 'text') { %>
-            .<%= fields[idx].fieldName %>ContentType(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase()%>_CONTENT_TYPE)<% } %><% } %>;
+            .<%= fields[idx].fieldName %>(<%=fieldStatus + fields[idx].fieldNameUnderscored.toUpperCase()%>)<% if ((fields[idx].fieldType === 'byte[]' || fields[idx].fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent !== 'text') { %>
+            .<%= fields[idx].fieldName %>ContentType(<%=fieldStatus + fields[idx].fieldNameUnderscored.toUpperCase()%>_CONTENT_TYPE)<% } %><% } %>;
         <%_ } else { _%>
         <%= asEntity(entityClass) %> <%= asEntity(entityInstance) %> = new <%= asEntity(entityClass) %>();
             <%_ for (idx in fields) { _%>
-        <%= asEntity(entityInstance) %>.set<%= fields[idx].fieldInJavaBeanMethod %>(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase() %>);
+        <%= asEntity(entityInstance) %>.set<%= fields[idx].fieldInJavaBeanMethod %>(<%=fieldStatus + fields[idx].fieldNameUnderscored.toUpperCase() %>);
                 <%_ if ((fields[idx].fieldType === 'byte[]' || fields[idx].fieldType === 'ByteBuffer') && fields[idx].fieldTypeBlobContent !== 'text') { _%>
-        <%= asEntity(entityInstance) %>.set<%= fields[idx].fieldInJavaBeanMethod %>ContentType(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase() %>_CONTENT_TYPE);
+        <%= asEntity(entityInstance) %>.set<%= fields[idx].fieldInJavaBeanMethod %>ContentType(<%=fieldStatus + fields[idx].fieldNameUnderscored.toUpperCase() %>_CONTENT_TYPE);
                 <%_ } _%>
             <%_ } _%>
         <%_ } _%>
@@ -442,6 +443,7 @@ _%>
             if ((relationshipValidate !== null && relationshipValidate === true) || mapsIdUse === true) { _%>
         // Add required entity
             <%_ if (alreadyGeneratedEntities.indexOf(otherEntityName) == -1) { _%>
+    <%_ if (otherEntityName === 'user') { // TODO or other entity has no unique fields _%>
         <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity(<% if (databaseType === 'sql') { %>em<% } %>);
                 <%_ if (databaseType === 'sql') { _%>
         em.persist(<%= otherEntityName %>);
@@ -450,6 +452,21 @@ _%>
                 <%_ if (databaseType === 'mongodb') { _%>
         <%= otherEntityName %>.setId("fixed-id-for-tests");
                 <%_ } _%>
+    <%_ } else { _%>
+        <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %>;
+        if (TestUtil.findAll(em, <%= otherEntityNameCapitalized %>.class).isEmpty()) {
+            <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.create<% if (fieldStatus === 'UPDATED_') { %>Updated<% } %>Entity(<% if (databaseType === 'sql') { %>em<% } %>);
+                <%_ if (databaseType === 'sql') { _%>
+            em.persist(<%= otherEntityName %>);
+            em.flush();
+                <%_ } _%>
+                <%_ if (databaseType === 'mongodb') { _%>
+            <%= otherEntityName %>.setId("fixed-id-for-tests");
+                <%_ } _%>
+        } else {
+            <%= otherEntityName %> = TestUtil.findAll(em, <%= otherEntityNameCapitalized %>.class).get(0);
+        }
+    <%_ } _%>
             <%_ } _%>
             <%_ if (relationshipType === 'many-to-many' || relationshipType === 'one-to-many') { _%>
         <%= asEntity(entityInstance) %>.get<%= relationshipNameCapitalizedPlural %>().add(<%= otherEntityName %>);
@@ -460,6 +477,7 @@ _%>
         <%_ } } _%>
         return <%= asEntity(entityInstance) %>;
     }
+<%_ }); _%>
 
     @BeforeEach
     public void initTest() {
@@ -560,7 +578,7 @@ _%>
             if (mapsIdUse === true) { _%>
         // Add a new parent entity
                 <%_ if (alreadyGeneratedEntities.indexOf(otherEntityName) == -1) { _%>
-        <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.createEntity(em);
+        <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %> = <%= otherEntityNameCapitalized %>ResourceIT.create<% if (otherEntityName !== 'user') { _%>Updated<%_ } %>Entity(em);
         em.persist(<%= otherEntityName %>);
         em.flush();
                 <%_ } _%>
@@ -813,6 +831,10 @@ _%>
     @Test<% if (databaseType === 'sql') { %>
     @Transactional<% } %>
     public void getAll<%= entityClassPlural %>By<%= relationship.relationshipNameCapitalized %>IsEqualToSomething() throws Exception {
+        <%_ if (relationship.relationshipValidate === true || relationship.useJPADerivedIdentifier === true) { _%>
+        // Get already existing entity
+        <%= asEntity(relationship.otherEntityNameCapitalized) %> <%= relationship.relationshipFieldName %> = <%= asEntity(entityInstance) %>.get<%= relationship.relationshipNameCapitalized %>();
+        <%_ } else { _%>
         // Initialize the database
         <%= asEntity(relationship.otherEntityNameCapitalized) %> <%= relationship.relationshipFieldName %> = <%= relationship.otherEntityNameCapitalized %>ResourceIT.createEntity(em);
         em.persist(<%= relationship.relationshipFieldName %>);
@@ -825,6 +847,7 @@ _%>
         <%= relationship.relationshipFieldName %>.set<%= relationship.otherEntityRelationshipNameCapitalized %>(<%= asEntity(entityInstance) %>);
 <%_     } _%>
 <%_ } _%>
+        <%_ } _%>
         <%= entityInstance %>Repository.saveAndFlush(<%= asEntity(entityInstance) %>);
         <% if (authenticationType === 'oauth2' && relationship.relationshipFieldName === 'user') { _%>String<%_ } else { _%>Long<% } %> <%= relationship.relationshipFieldName %>Id = <%= relationship.relationshipFieldName %>.getId();
 

--- a/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
+++ b/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
@@ -35,6 +35,7 @@ import org.springframework.security.test.context.TestSecurityContextHolder;
 import java.io.IOException;
 import java.time.ZonedDateTime;
 import java.time.format.DateTimeParseException;
+<%_ if (databaseType === 'sql') { _%>
 import java.util.List;
 
 import javax.persistence.EntityManager;
@@ -42,6 +43,7 @@ import javax.persistence.TypedQuery;
 import javax.persistence.criteria.CriteriaBuilder;
 import javax.persistence.criteria.CriteriaQuery;
 import javax.persistence.criteria.Root;
+<%_ } _%>
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -171,6 +173,7 @@ public final class TestUtil {
         return dfcs;
     }
 
+<%_ if (databaseType === 'sql') { _%>
     /**
      * Makes a an executes a query to the EntityManager finding all stored objects.
      * @param <T> The type of objects to be searched
@@ -186,6 +189,7 @@ public final class TestUtil {
         TypedQuery<T> allQuery = em.createQuery(all);
         return allQuery.getResultList();
     }
+<%_ } _%>
 
     private TestUtil() {}
 }

--- a/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
+++ b/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
@@ -179,7 +179,7 @@ public final class TestUtil {
      * @param <T> The type of objects to be searched
      * @param em The instance of the EntityManager
      * @param clss The class type to be searched
-     * @return A list of all found objects of the
+     * @return A list of all found objects
      */
     public static <T> List<T> findAll(EntityManager em, Class<T> clss) {
         CriteriaBuilder cb = em.getCriteriaBuilder();

--- a/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
+++ b/generators/server/templates/src/test/java/package/web/rest/TestUtil.java.ejs
@@ -35,6 +35,13 @@ import org.springframework.security.test.context.TestSecurityContextHolder;
 import java.io.IOException;
 import java.time.ZonedDateTime;
 import java.time.format.DateTimeParseException;
+import java.util.List;
+
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -162,6 +169,22 @@ public final class TestUtil {
         registrar.setUseIsoFormat(true);
         registrar.registerFormatters(dfcs);
         return dfcs;
+    }
+
+    /**
+     * Makes a an executes a query to the EntityManager finding all stored objects.
+     * @param <T> The type of objects to be searched
+     * @param em The instance of the EntityManager
+     * @param clss The class type to be searched
+     * @return A list of all found objects of the
+     */
+    public static <T> List<T> findAll(EntityManager em, Class<T> clss) {
+        CriteriaBuilder cb = em.getCriteriaBuilder();
+        CriteriaQuery<T> cq = cb.createQuery(clss);
+        Root<T> rootEntry = cq.from(clss);
+        CriteriaQuery<T> all = cq.select(rootEntry);
+        TypedQuery<T> allQuery = em.createQuery(all);
+        return allQuery.getResultList();
     }
 
     private TestUtil() {}

--- a/test-integration/samples/jdl-default/app.jdl
+++ b/test-integration/samples/jdl-default/app.jdl
@@ -105,6 +105,18 @@ entity Task {
   description String
 }
 
+entity GoldenBadge {
+  name String
+}
+
+entity SilverBadge {
+  name String
+}
+
+entity Identifier {
+  name String required unique
+}
+
 entity Country {
   name String
 }
@@ -114,7 +126,8 @@ entity Region {
 }
 
 relationship OneToOne {
-  Department{location} to Location
+  Department{location} to Location,
+  Employee{user(login)} to User with jpaDerivedIdentifier
 }
 
 relationship OneToMany {
@@ -138,25 +151,28 @@ relationship OneToMany {
   Employee{department},
   Employee{job} to Job{emp(lastName)},
   Location{country} to Country,
-  Country{area(region)} to Region
+  Country{area(name)} to Region
 }
 
 relationship ManyToOne {
-  Employee{user(login)} to User{employee},
-  Employee{manager(lastName)} to Employee
+  Employee{manager(lastName)} to Employee,
+  Employee{sibag(name) required} to SilverBadge,
+  Employee{gobag(name) required} to GoldenBadge,
+  SilverBadge{iden(name) required} to Identifier,
+  GoldenBadge{iden(name) required} to Identifier
 }
 
 relationship ManyToMany {
   JobHistory to Department,
   JobHistory to Job{history},
-  JobHistory{emp(employee)} to Employee{history},
-  Job{chore(title)} to Task{linkedJob(JobTitle)}
+  JobHistory{emp(firstName)} to Employee{history},
+  Job{chore(title)} to Task{linkedJob(title)}
 }
 
-dto BankAccount, Employee, Department, Location, Country, Region with mapstruct
+dto BankAccount, Employee, Department, Location, Country, Region, SilverBadge, GoldenBadge, Identifier with mapstruct
 
 angularSuffix BankAccount with mySuffix
-filter BankAccount
+filter BankAccount, Employee
 clientRootFolder BankAccount, TheLabel, Operation with test-root
 
 paginate TheLabel, Job with pagination

